// @solid-primitives/scroll@2.0.20/dist/index.cjs
'use strict';

var eventListener = require('@solid-primitives/event-listener');
var rootless = require('@solid-primitives/rootless');
var staticStore = require('@solid-primitives/static-store');
var solidJs = require('solid-js');
var web = require('solid-js/web');

// src/index.ts
function getScrollParent(node) {
  if (web.isServer) {
    return {};
  }
  while (node && !isScrollable(node)) {
    node = node.parentElement;
  }
  return node || document.scrollingElement || document.documentElement;
}
function isScrollable(node) {
  if (web.isServer) {
    return false;
  }
  const style = window.getComputedStyle(node);
  return /(auto|scroll)/.test(style.overflow + style.overflowX + style.overflowY);
}
var FALLBACK_SCROLL_POSITION = { x: 0, y: 0 };
function getScrollPosition(target) {
  if (web.isServer || !target) {
    return { ...FALLBACK_SCROLL_POSITION };
  }
  if (target instanceof Window)
    return {
      x: target.scrollX,
      y: target.scrollY
    };
  return {
    x: target.scrollLeft,
    y: target.scrollTop
  };
}
function createScrollPosition(target) {
  if (web.isServer) {
    return FALLBACK_SCROLL_POSITION;
  }
  target = target || window;
  const isFn = typeof target === "function", isHydrating = solidJs.sharedConfig.context, getTargetPos = isFn ? () => getScrollPosition(target()) : () => getScrollPosition(target), [calc, setCalc] = solidJs.createSignal(isHydrating ? () => FALLBACK_SCROLL_POSITION : getTargetPos, {
    equals: false
  }), trigger = () => setCalc(() => getTargetPos), pos = staticStore.createDerivedStaticStore(() => calc()());
  if (isHydrating || isFn)
    solidJs.onMount(trigger);
  eventListener.createEventListener(target, "scroll", trigger, { passive: true });
  return pos;
}
exports.useWindowScrollPosition = /* @__PURE__ */ rootless.createHydratableSingletonRoot(
  () => createScrollPosition(web.isServer ? () => void 0 : window)
);

exports.createScrollPosition = createScrollPosition;
exports.getScrollParent = getScrollParent;
exports.getScrollPosition = getScrollPosition;
exports.isScrollable = isScrollable;
