// @solid-primitives/scroll@2.0.20/dist/index.js
import { createEventListener } from '@solid-primitives/event-listener';
import { createHydratableSingletonRoot } from '@solid-primitives/rootless';
import { createDerivedStaticStore } from '@solid-primitives/static-store';
import { sharedConfig, createSignal, onMount } from 'solid-js';
import { isServer } from 'solid-js/web';

// src/index.ts
function getScrollParent(node) {
  if (isServer) {
    return {};
  }
  while (node && !isScrollable(node)) {
    node = node.parentElement;
  }
  return node || document.scrollingElement || document.documentElement;
}
function isScrollable(node) {
  if (isServer) {
    return false;
  }
  const style = window.getComputedStyle(node);
  return /(auto|scroll)/.test(style.overflow + style.overflowX + style.overflowY);
}
var FALLBACK_SCROLL_POSITION = { x: 0, y: 0 };
function getScrollPosition(target) {
  if (isServer || !target) {
    return { ...FALLBACK_SCROLL_POSITION };
  }
  if (target instanceof Window)
    return {
      x: target.scrollX,
      y: target.scrollY
    };
  return {
    x: target.scrollLeft,
    y: target.scrollTop
  };
}
function createScrollPosition(target) {
  if (isServer) {
    return FALLBACK_SCROLL_POSITION;
  }
  target = target || window;
  const isFn = typeof target === "function", isHydrating = sharedConfig.context, getTargetPos = isFn ? () => getScrollPosition(target()) : () => getScrollPosition(target), [calc, setCalc] = createSignal(isHydrating ? () => FALLBACK_SCROLL_POSITION : getTargetPos, {
    equals: false
  }), trigger = () => setCalc(() => getTargetPos), pos = createDerivedStaticStore(() => calc()());
  if (isHydrating || isFn)
    onMount(trigger);
  createEventListener(target, "scroll", trigger, { passive: true });
  return pos;
}
var useWindowScrollPosition = /* @__PURE__ */ createHydratableSingletonRoot(
  () => createScrollPosition(isServer ? () => void 0 : window)
);

export { createScrollPosition, getScrollParent, getScrollPosition, isScrollable, useWindowScrollPosition };
